server/src/main/java/org/elasticsearch/index/fieldvisitor/CustomFieldsVisitor.java

```diff
@@ -10,21 +10,24 @@
 import org.apache.lucene.index.FieldInfo;
 
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 
 /**
- * A field visitor that allows to load a selection of the stored fields by exact name or by pattern.
- * Supported pattern styles: "xxx*", "*xxx", "*xxx*" and "xxx*yyy".
- * The Uid field is always loaded.
- * The class is optimized for source loading as it is a common use case.
+ * A field visitor that allows to load a selection of the stored fields by exact name
+ * {@code _id} and {@code _routing} fields are always loaded.
  */
 public class CustomFieldsVisitor extends FieldsVisitor {
 
     private final Set<String> fields;
 
     public CustomFieldsVisitor(Set<String> fields, boolean loadSource) {
         super(loadSource);
-        this.fields = fields;
+        this.fields = new HashSet<>(fields);
+        // metadata fields are already handled by FieldsVisitor, so removing
+        // them here means that if the only fields requested are metadata
+        // fields then we can shortcut loading
+        List.of("_id", "_routing", "_source").forEach(this.fields::remove);
```

Comment:
It's a bit tricksy, because we need some way of distinguishing 'load metadata but don't load anything else' from 'don't load anything at all'.  I *could* add another boolean to StoredFieldsSpec called `loadMetadata` I suppose?  Maybe I should give that a go and see if it's less obviously weird.

I think it's still worth removing these fields from the list though, because users can get confused and ask for them as part of a stored fields list.  But maybe that's enough of a corner case that it doesn't matter?

This is all convincing me that we really want to merge loading stored fields into ValueFetchers (which this will make possible!) and deprecating `stored_fields`, `docvalue_fields` and `script_fields` altogether.

Reply:
> deprecating stored_fields, docvalue_fields and script_fields altogether.

:+1:

Do we need to remove these fields because sometimes folks will write them into `stored_fields` or `fields` or whatever and we have to respect that, but we do need to understand that they are metadata at this point? I'm ok with keeping the removal. I was just wondering if there should be a next step to make this less confusing.

Reply:
Right, if they add them in it has no effect on the final output (they get returned anyway because they're metadata fields and we always load metadata fields if stored fields or source are requested), but it can slow things down in certain circumstances, like if you've explicitly said not to load source we end up loading it and then ignoring it.